### PR TITLE
feat(backend): contact form send from backend to brevo

### LINKED ISSUE #51: 🚀 [Feature] Contact Form send from backend to Brevo
## 🚀 Feature

Contact Form send from backend to Brevo

image

Data:
- Vorname
- nachname
- email
- frage (content)
- (dsvgo)

Goal:
- Send Static Data in the given Form from our backend to brevo.
- Receive two emails:
  - One for the admin/support:
    - Daten aus Kontaktformular an uns (template1)
  - one for the Questioneer:
    - obige ummantelt mit Info, dass diese Daten an die DramMall E-Mail versendet wurden,
an Nutzer-E-Mail-Adresse (template2)

### PR DESCRIPTION
**Pull request by Elweyn**
### Issues

- fixes #51 

### Todo

- [X] Send Email to given contact data
- [x] Send Email to Admin
- [x] Configure Admin in env
- [x] Write Unit Tests

### CODE REVIEW

File: backend/.env.dist [hunk 1]
```diff
@@ -1 +1,2 @@
+BREVO_KEY="123456"
```
Comment (ulfgebhardt) [inline backend/.env.dist:1]: should be empty by default - which will then disable the whole service

File: backend/src/api/NewsletterBrevo.ts [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+import { NewsletterEmailAttributes } from '#model/NewsletterEmailAttributes'
+
+export const createBrevoInstance = (): SibApiV3Sdk.TransactionalEmailsApi => {
+  const apiInstance = new SibApiV3Sdk.TransactionalEmailsApi()
+  apiInstance.setApiKey(SibApiV3Sdk.TransactionalEmailsApiApiKeys.apiKey, config.BREVO_KEY)
+  return apiInstance
+}
+
+export const createSmtpEmail = (
+  emailAttributes: NewsletterEmailAttributes,
```
Comment (ulfgebhardt) [inline backend/src/api/NewsletterBrevo.ts:14]: Why not use the types of the library but define your own? Why is this needed?

File: backend/src/config/config.ts [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+// Load env file
+config({
+  path: path.resolve(__dirname, '../../.env'),
+})
+
+interface ENV {
+  BREVO_KEY: string | undefined
+}
+
+interface Config {
```
Comment (ulfgebhardt) [inline backend/src/config/config.ts:14]: can't that be the same?

File: backend/src/graphql/resolvers/ContactFormResolver.ts [hunk 4]
```diff
@@ -11,6 +14,43 @@ export class ContactFormResolver {
...
+          email: contactFormData.email,
+        },
+      ],
+      htmlContent:
+        '<html><body><h1>This is my first transactional email {{params.parameter}}</h1></body></html>',
+      sender: { name: 'DreamMall Earth Team', email: 'no-reply@dreammall.earth' },
+      bcc: [
+        { name: 'Mathias Lenz', email: 'mathias.lenz@it4c.dev' },
+        { name: 'Hannes Heine', email: 'hannes.heine@it4c.dev' },
+        { name: 'Ulf Gebhardt', email: 'ulf.gebhardt@it4c.dev' },
```
Comment (ulfgebhardt) [inline backend/src/graphql/resolvers/ContactFormResolver.ts:33]: i do not want my email publicly on github pls

File: backend/src/graphql/resolvers/ContactFormResolver.ts [hunk 5]
```diff
@@ -11,6 +14,43 @@ export class ContactFormResolver {
...
+    const sendSmtpEmail: SibApiV3Sdk.SendSmtpEmail = createSmtpEmail({
+      subject: 'My {{params.subject}}',
+      emailTo: [
+        {
+          name: contactFormData.firstName + ' ' + contactFormData.lastName,
+          email: contactFormData.email,
+        },
+      ],
+      htmlContent:
+        '<html><body><h1>This is my first transactional email {{params.parameter}}</h1></body></html>',
```
Comment (ulfgebhardt) [inline backend/src/graphql/resolvers/ContactFormResolver.ts:28]: we want to use a template managed by brevo here

File: backend/src/model/NewsletterEmailAttributes.ts [hunk 6]
```diff
@@ -0,0 +1,15 @@
+import { NewsletterHeader } from './NewsletterHeader'
```
Comment (ulfgebhardt) [inline backend/src/model/NewsletterEmailAttributes.ts:1]: I do not see the point of this and the naming is not on point either.

This is neither a graphQL-Model nor a database-Model - its just some interface classes/types.

Please explain to me the purpose of this whole folder? Shouldnt it just be parameters to the corresponding function?

File: backend/src/graphql/resolvers/ContactFormResolver.ts [hunk 7]
```diff
@@ -11,6 +14,43 @@ export class ContactFormResolver {
   ): Promise<boolean> {
     await prisma.contactForm.create({ data: contactFormData })
     // code to send email goes here
+    const apiInstance: SibApiV3Sdk.TransactionalEmailsApi = createBrevoInstance()
+
+    const sendSmtpEmail: SibApiV3Sdk.SendSmtpEmail = createSmtpEmail({
+      subject: 'My {{params.subject}}',
```
Comment (ulfgebhardt) [inline backend/src/graphql/resolvers/ContactFormResolver.ts:20]: this might be part of a brevo template

File: backend/src/graphql/resolvers/ContactFormResolver.ts (hunk 5 above)
Comment (ulfgebhardt) [reply backend/src/graphql/resolvers/ContactFormResolver.ts:28]: image

That way you can just pass the whole object from the api to brevo and that handles all the rest.
https://developers.brevo.com/docs/send-a-transactional-email#1-generate-a-code-snippet-to-quickly-test-your-request

File: backend/tsconfig.json [hunk 8]
```diff
@@ -8,8 +8,11 @@
     "baseUrl": ".",
     "lib": ["ESNext"],
     "paths": {
+      "#api/*": ["./src/api/*"],
+      "#config/*": ["./src/config/*"],
       "#graphql/*": ["./src/graphql/*"],
       "#inputs/*": ["./src/graphql/inputs/*"],
+      "#model/*": ["./src/model/*"],
```
Comment (Mogge) [inline backend/tsconfig.json:15]: You need this in the packet.json as well

File: backend/src/graphql/resolvers/ContactFormResolver.ts [hunk 9]
```diff
@@ -11,7 +14,10 @@ export class ContactFormResolver {
   ): Promise<boolean> {
     await prisma.contactForm.create({ data: contactFormData })
     // code to send email goes here
-    return true
+    const apiInstance: SibApiV3Sdk.TransactionalEmailsApi = createBrevoInstance()
```
Comment (Mogge) [inline backend/src/graphql/resolvers/ContactFormResolver.ts:17]: I suggest to import a void function here, that takes the DB entry created above as argument, and handles all the emails sending. This function can update the db entry when sending the email was successful.

File: backend/src/graphql/resolvers/ContactFormResolver.ts [hunk 10]
```diff
@@ -11,7 +14,10 @@ export class ContactFormResolver {
...
+
+    const success: boolean = createSmtpEmail(apiInstance, contactFormData)
+    return success
```
Comment (Mogge) [inline backend/src/graphql/resolvers/ContactFormResolver.ts:20]: I suggest to return true, independent of the email. We do not want to wait for a response for sending the email.

File: backend/src/api/NewsletterBrevo.ts [hunk 11]
```diff
@@ -0,0 +1,47 @@
+// eslint-disable-next-line import/no-namespace
+import * as SibApiV3Sdk from '@getbrevo/brevo'
+
+import config from '#config/config'
+import { ContactFormInput } from '#graphql/inputs/ContactFormInput'
```
Comment (Mogge) [inline backend/src/api/NewsletterBrevo.ts:5]:
```suggestion
import { ContactForm } from '@prisma/client'
```

File: backend/src/api/NewsletterBrevo.ts [hunk 12]
```diff
@@ -0,0 +1,47 @@
...
+
+export const createBrevoInstance = (): SibApiV3Sdk.TransactionalEmailsApi => {
+  const apiInstance = new SibApiV3Sdk.TransactionalEmailsApi()
+  apiInstance.setApiKey(SibApiV3Sdk.TransactionalEmailsApiApiKeys.apiKey, config.BREVO_KEY)
+  return apiInstance
+}
+
+export const createSmtpEmail = (
+  apiInstance: SibApiV3Sdk.TransactionalEmailsApi,
+  contactFormData: ContactFormInput,
```
Comment (Mogge) [inline backend/src/api/NewsletterBrevo.ts:15]:
```suggestion
  contactFormData: ContactForm,
```
This way you have the id, and you can update the db entry when sending the email was successful

File: backend/src/graphql/resolvers/ContactFormResolver.ts [hunk 13]
```diff
@@ -11,7 +14,10 @@ export class ContactFormResolver {
   ): Promise<boolean> {
     await prisma.contactForm.create({ data: contactFormData })
```
Comment (Mogge) [inline backend/src/graphql/resolvers/ContactFormResolver.ts:15]:
```suggestion
    const contactForm = await prisma.contactForm.create({ data: contactFormData })
```

File: backend/src/graphql/resolvers/ContactFormResolver.ts [hunk 14]
```diff
@@ -11,7 +14,10 @@ export class ContactFormResolver {
...
     // code to send email goes here
-    return true
+    const apiInstance: SibApiV3Sdk.TransactionalEmailsApi = createBrevoInstance()
+
+    const success: boolean = createSmtpEmail(apiInstance, contactFormData)
```
Comment (Mogge) [inline backend/src/graphql/resolvers/ContactFormResolver.ts:19]:
```suggestion
    const success: boolean = createSmtpEmail(apiInstance, contactForm)
```

**Review by Mogge — COMMENTED** [review]
Some suggestions to discuss

File: backend/src/graphql/resolvers/ContactFormResolver.ts (hunk 10 above)
Comment (Elweyn) [reply backend/src/graphql/resolvers/ContactFormResolver.ts:20]: Good point!

File: backend/tsconfig.json (hunk 8 above)
Comment (Elweyn) [reply backend/tsconfig.json:15]: This is not needed like @ulfgebhardt pointed out we can use the Brevo objects!

File: backend/src/graphql/resolvers/ContactFormResolver.ts (hunk 4 above)
Comment (mahula) [reply backend/src/graphql/resolvers/ContactFormResolver.ts]: Why a bcc list?
Contact form submission should trigger just two e-mails
- to info/contact@dreammall.earth
- to e-mail address provided by the user in the contact form

Comment (Elweyn) [reply backend/src/graphql/resolvers/ContactFormResolver.ts]: Good point!

File: backend/src/graphql/resolvers/ContactFormResolver.ts (hunk 5 above)
Comment (Elweyn) [reply backend/src/graphql/resolvers/ContactFormResolver.ts]: Yes, wanted to have a email that goes out first but you are right we don't want the form that is implemented yet, but we want the templateId solution!

File: backend/src/config/config.ts (hunk 3 above)
Comment (Elweyn) [reply backend/src/config/config.ts:14]: ENV defines that it could not be present in .env & Config has the value given after it and is tested in getSanitzedConfig.
So if the .env file has not the value BREVO_KEY in it empty or filled we get an error

File: backend/src/config/config.ts [hunk 15]
```diff
@@ -0,0 +1,44 @@
...
+  BREVO_KEY: string | undefined
+  BREVO_CONTACT_REQUEST_TO_NAME: string | undefined
+  BREVO_CONTACT_REQUEST_TO_EMAIL: string | undefined
+}
+
+interface Config {
+  BREVO_KEY: string
+  BREVO_CONTACT_REQUEST_TO_NAME: string
+  BREVO_CONTACT_REQUEST_TO_EMAIL: string
+}
```
Comment (ulfgebhardt) [inline backend/src/config/config.ts:20]: i still think this should be the same

and below you should not throw an error - this breaks the unittests?!

File: backend/src/graphql/resolvers/ContactFormResolver.ts [hunk 16]
```diff
@@ -9,8 +14,57 @@ export class ContactFormResolver {
   async createContactForm(
     @Arg('contactFormData') contactFormData: ContactFormInput,
   ): Promise<boolean> {
-    await prisma.contactForm.create({ data: contactFormData })
+    const contactForm: ContactForm = await prisma.contactForm.create({ data: contactFormData })
     // code to send email goes here
+    const smtpEmailToAdmin: SibApiV3Sdk.SendSmtpEmail = createSmtpEmail(
+      1,
+      [
```
Comment (ulfgebhardt) [inline backend/src/graphql/resolvers/ContactFormResolver.ts:21]: use a constant for template id

File: backend/src/graphql/resolvers/ContactFormResolver.ts [hunk 17]
```diff
@@ -9,8 +14,57 @@ export class ContactFormResolver {
   async createContactForm(
     @Arg('contactFormData') contactFormData: ContactFormInput,
   ): Promise<boolean> {
-    await prisma.contactForm.create({ data: contactFormData })
+    const contactForm: ContactForm = await prisma.contactForm.create({ data: contactFormData })
     // code to send email goes here
+    const smtpEmailToAdmin: SibApiV3Sdk.SendSmtpEmail = createSmtpEmail(
```
Comment (ulfgebhardt) [inline backend/src/graphql/resolvers/ContactFormResolver.ts:19]: you should encapsulate all this into a function call like seen in line 17 with the database and have all this clutter live in `api/NewsletterBrevo.ts` - there you expose one function `export sendContactFormEmail` which takes your data and handles everything

File: backend/src/config/config.ts (hunk 3 above)
Comment (ulfgebhardt) [reply backend/src/config/config.ts]: why? - that breaks your unit testing

Comment (ulfgebhardt) [reply backend/src/config/config.ts]: image

File: backend/src/graphql/resolvers/ContactFormResolver.ts (hunk 16 above)
Comment (ulfgebhardt) [reply backend/src/graphql/resolvers/ContactFormResolver.ts:21]: maybe also configurable via config since could be required to change easily

File: backend/src/config/config.ts (hunk 3 above)
Comment (Elweyn) [reply backend/src/config/config.ts]: Because I want to know if I setup the server and don't have some keys in my .env that they are missing!

File: backend/src/config/config.ts (hunk 15 above)
Comment (Elweyn) [reply backend/src/config/config.ts]: I understand what you mean but this allows us to see the missing keys when we use:
```bash
# backend
npm run start
npm run dev
```

File: backend/test/testSetup.ts [hunk 18]
```diff
@@ -1,2 +1,6 @@
 // eslint-disable-next-line import/no-unassigned-import
 import 'reflect-metadata'
+
+const consoleWarnMock = jest.fn()
+// eslint-disable-next-line no-console
+console.warn = consoleWarnMock
```
Comment (Mogge) [inline backend/test/testSetup.ts:6]: Please no. I think even we should do something like
`delete global.console.warn
global.console.warn = global.console.error
```

File: backend/src/config/config.ts [hunk 19]
```diff
@@ -0,0 +1,55 @@
...
+  BREVO_TEMPLATE_CONTACT_USER: number | undefined
+}
+
+interface Config {
+  BREVO_KEY: string
+  BREVO_CONTACT_REQUEST_TO_NAME: string
+  BREVO_CONTACT_REQUEST_TO_EMAIL: string
+  BREVO_TEMPLATE_CONTACT_BASE: number
+  BREVO_TEMPLATE_CONTACT_USER: number
+}
```
Comment (ulfgebhardt) [inline backend/src/config/config.ts:24]: this is still an issue

File: backend/src/api/NewsletterBrevo.ts [hunk 20]
```diff
@@ -0,0 +1,109 @@
...
+    },
+    // eslint-disable-next-line promise/prefer-await-to-callbacks
+    function (error) {
+      // eslint-disable-next-line no-console
+      console.error(error)
+    },
+  )
+}
+
+export const sendContactFormEmail = (contactForm: ContactForm): void => {
```
Comment (ulfgebhardt) [inline backend/src/api/NewsletterBrevo.ts:58]: promise handling is not correct -> you should always return promises and explicitly ignore them on the top level (api resolver)

Comment (ulfgebhardt) [reply backend/src/api/NewsletterBrevo.ts:58]: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise/all

File: backend/src/config/config.ts (hunk 19 above)
Comment (Elweyn) [reply backend/src/config/config.ts:24]: No to have the behavior of the console warn we need both!
- If I change ENV to Config => I get the 
Screenshot 2023-12-18 at 16 13 30

- If I change Config to ENV => I get the same on 
Screenshot 2023-12-18 at 16 14 41

File: backend/src/api/NewsletterBrevo.ts (hunk 20 above)
Comment (Elweyn) [reply backend/src/api/NewsletterBrevo.ts:58]: OK So instead of updating the Database in here, I do it on top level?
So I should return the function as promise?

Let speak about it so I am sure to understand it right
